webapp/src/main/java/com/box/l10n/mojito/entity/ApplicationCache.java

```diff
@@ -0,0 +1,83 @@
+package com.box.l10n.mojito.entity;
+
+import org.hibernate.annotations.BatchSize;
+import org.hibernate.annotations.Type;
+import org.joda.time.DateTime;
+
+import javax.persistence.Column;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.Index;
+import javax.persistence.Lob;
+import javax.persistence.Table;
+
+/**
+ * Entity that contains the cache entry details for a database-backed application cache.
+ * Each entity is identified by a specific {@Link ApplicationCacheType} and MD5 hash of the key.
+ * A TTL / expiry date can also be configured on the entries, the application being responsible for the entry eviction.
+ *
+ * @author garion
+ */
+@Entity
+@Table(name = "application_cache",
+        indexes = {
+                @Index(name = "I__APPLICATION_CACHE__CACHE_TYPE_ID_EXPIRY_DATE", columnList = "cache_type_id, expiry_date"),
+        })
+@BatchSize(size = 1000)
+public class ApplicationCache {
+    @EmbeddedId
```

Comment:
I followed the best practices defined here for the composite key setup: https://vladmihalcea.com/the-best-way-to-map-a-composite-primary-key-with-jpa-and-hibernate/ 

I generally use a surrogate PK key for most entities (to help with joins performance etc.) - but in this case a natural key seems a better fit - as this is a completely self contained table - and an auto-incremented ID PK not only doesn't give us anything extra - but it makes the lookups slightly slower.

As we primarily lookup by the cache id + hash of the key (and we would never lookup by an auto-incremented ID for a cache) - having these as a natural key used for clustering data is likely close to the ideal setup.

That said - I am interested in learning more about the best practices you mentioned - can you expand on what should be the ideal for this specific scenario?

Reply:
I don't remember exactly at this point... more of a vague memory. Not sure if it was an hibernate thing (the embedded id maybe have some complication with implementing serializable, or maybe it can lead to doing a lot of overcomplicated things) or maybe more of general DB design recommandation (as you said having surrogate PK just make it simpler for join and understanding things quickly). Whatever I remember might be outdated too.

Reply:
ooopss I meant sticking with simple id. A bit late now, but DB schema is hard to change later. Are you sure about the benefis for the composed key?